xiilib/_gunicorn/charm.py

```diff
@@ -0,0 +1,191 @@
+        """
+        super().__init__(framework)
+        self._secret_storage = GunicornSecretStorage(
+            charm=self, key=f"{wsgi_framework}_secret_key"
+        )
+        self._database_requirers = make_database_requirers(self, self.app.name)
+        try:
+            wsgi_config = self.get_wsgi_config()
+        except CharmConfigInvalidError as exc:
+            self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
+            return
+
+        self._charm_state = CharmState.from_charm(
+            charm=self,
+            framework=wsgi_framework,
+            wsgi_config=wsgi_config,
+            secret_storage=self._secret_storage,
+            database_requirers=self._database_requirers,
+        )
+        self._database_migration = DatabaseMigration(
+            container=self.unit.get_container(self._charm_state.container_name),
+            state_dir=self._charm_state.state_dir,
+        )
+        webserver = GunicornWebserver(
+            charm_state=self._charm_state,
+            container=self.unit.get_container(self._charm_state.container_name),
+        )
+        self._wsgi_app = WsgiApp(
+            container=self.unit.get_container(f"{self._charm_state.framework}-app"),
+            charm_state=self._charm_state,
+            webserver=webserver,
+            database_migration=self._database_migration,
+        )
+        self._databases = Databases(
+            charm=self,
+            application=self._wsgi_app,
+            database_requirers=self._database_requirers,
+        )
+        self._ingress = IngressPerAppRequirer(
+            self,
+            port=self._charm_state.port,
+            strip_prefix=True,
+        )
+        self._observability = Observability(
+            self,
+            charm_state=self._charm_state,
+            container_name=self._charm_state.container_name,
+            cos_dir=self.get_cos_dir(),
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.rotate_secret_key_action, self._on_rotate_secret_key_action)
+        self.framework.observe(
+            self.on.secret_storage_relation_changed,
+            self._on_secret_storage_relation_changed,
+        )
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        for database, database_requirer in self._database_requirers.items():
+            self.framework.observe(
+                database_requirer.on.database_created,
+                getattr(self, f"_on_{database}_database_database_created"),
+            )
+            self.framework.observe(
+                database_requirer.on.endpoints_changed,
+                getattr(self, f"_on_{database}_database_endpoints_changed"),
+            )
+            self.framework.observe(
+                self.on[database_requirer.relation_name].relation_broken,
+                getattr(self, f"_on_{database}_database_relation_broken"),
+            )
```

Comment:
Why don't we apply our observer pattern here, and use a database observer ?

Reply:
This is originally an observer, and it has to be changed into this because of the circular dependency issues caused by the "Managing Charm Complexity"

Reply:
We also discussed this in another spec and, to enable customisation, the observe events should actually be on the charm and that means we can't really use observers too much. See ISD088. We should actually double check - I think there might be some observers here

Reply:
Ok the question is more, which events to we want to provide customizability. These ones could be tricky because of the database name being in the event handler name

Reply:
I think we should support customising all of the events, you also noted that there is a secret rotation event we should probably also move out to the charm